content/geometry/CircleIntersection.h

```diff
@@ -10,15 +11,12 @@
 #include "Point.h"
 
 typedef Point<double> P;
-bool circleIntersection(P a, P b, double r1, double r2,
-		pair<P, P>* out) {
+int circleInter(P a,P b,double r1,double r2,pair<P, P>* out){
+	if (a == b) { assert(r1 != r2); return false; }
 	P delta = b - a;
-	assert(delta.x || delta.y || r1 != r2);
-	if (!delta.x && !delta.y) return false;
-	double r = r1 + r2, d2 = delta.dist2();
-	double p = (d2 + r1*r1 - r2*r2) / (2.0 * d2);
+	double d2 = delta.dist2(), p = (d2 + r1*r1 - r2*r2)/(d2*2);
 	double h2 = r1*r1 - p*p*d2;
-	if (d2 > r*r || h2 < 0) return false;
+	if (h2 < 0) return false;
```

Comment:
Hm, so. This does lose some numeric precision; now circles that are disjoint might get an intersection due to bad rounding if we're not favored by the floating point gods. (We had similar problems in other cases even before.)

If I did my math right, a numerically stable way of writing this condition would be `(a+b+c)*(a+b-c)*(b-a+c)*(a-b+c) < 0`, where `a = r1`, `b = r2`, `c = sqrt(d2)` (of which the first term could be eliminated, and the rest probably reformulated in terms of inequalities). I wonder if the rest of the function could be written in terms of those variables...

Reply:
Well, I guess we could keep the rest of the function, but I think returning false if either `(r1 + r2)^2 < d2` or `(r1 - r2)^2 > d2`, and then letting `h2 = max(0.0, current expr);`, would be worthwhile